### ISSUE TITLE
trapread triggers up false positives

### ISSUE THREAD

**Issue by SvOlli** [stella-emu/stella#272]
In the demo Tricade by Trilobit (http://www.pouet.net/prod.php?which=32203), there is a hidden part mentioned. So I wanted to figure out, what triggers it. So I started stella in debug mode, and defined the following traps
```
trapread INPT0
trapread INPT1
trapread INPT2
trapread INPT3
trapread INPT4
trapread INPT5
trapread SWCHA
trapread SWCHB
```
After that stella breaks on a read on $93 and $EF12, 
(Edit: happens with current development version compiled from trunk,)

**Comment by thrust26** [comment]
Thanks for reporting. This is tricky, worth a discussion (BTW: 4.7.2 did not trap that way).

The trap results from the previous instruction `sta AUDV0, y` (Y=0). This is address $19. Which is a mirror of INPT1.

But this is a write, so why does the read trap trigger here? Because of the internals of the 650x CPU. Here is the definition:
```
define(M6502_ABSOLUTEY_WRITE, `{
  uInt16 low = peek(PC++, DISASM_CODE);
  uInt16 high = (uInt16(peek(PC++, DISASM_CODE)) << 8);
  peek(high | uInt8(low + Y), DISASM_NONE);  // <------ THIS!
  operandAddress = (high | low) + Y;
}')
```
As you can see, there is a dummy read in this (and many other addressing modes) and this triggers the trap. A fix would be to ignore dummy reads. But those could e.g. trigger hotspots, no? 

So what would be the expected behavior here?

**Comment by DirtyHairy** [comment]
I don't think we should do anything here. The 6502 does a bus access in every cycle, and the unwanted ones can be just as important as the others. I think this is a case of "works as designed" :smirk:

**Comment by thrust26** [comment]
Technically you are 100% right. But as you can see from this issue, it is irritating and surprising to even very advanced users like @SvOlli. 

Maybe we should offer a (global) debugger option to ignore such dummy peeks? I think I could handle that by just checking the flags. Even though they are meant to be used for DiStella, checking for DISASM_NONE should suit our needs her too.

**Comment by SvOlli** [comment]
I think a (more detailed) output on the debug console should also be worth a thought. Something like:
`trap id 00 triggered, due to access at $ef12 (mirror of ...)`

I'd prefer this to the global switch option, because the switch offers you a workaround for a bug that needs to be fixed in development.

I still remember the last time I stumbled over a ghost read triggering hardware: I had an RTS instruction at $FFF7 when running code on a SuperCharger, because I did not consider the ghost read at $FFF8, the main hotspot. Even though Stella's trap put my nose right on it, it took me several minutes to figure out, what was happening there. Having that switch, I'd most probably turned that one on, and looked for the problem somewhere else.

When analysing old code that triggers a ghost read, you can still include a hot-fix by patching the code.

**Comment by thrust26** [comment]
OK, I changed this into an enhancement.

**Comment by DirtyHairy** [comment]
I think it boils down to a matter of perspective. I usually use the debugger to identify the reason for an access that I know is happening, therefore the current behavior seems natural too me :smirk: On the other hand, if you use the debugger to find the locations where a register is intended to be read, the triggering on the bogus accesses is potentially undesired, I agree.

You've convinced me, I think a global option to toggle this makes sense, as does an extension of the console output 😈

**Comment by thrust26** [comment]
@SvOlli There was a bug in label name resolution which I fixed with 50c8c5c. This should improve the situation a bit.

And I have added the index of the condition causing the break too. Give it a try, please.

**Comment by sa666666** [comment]
I really have nothing else to add other than what has been said above.  I'd really like to have the output differentiate between intended and unintended reads, since this can be a major source of bugs in development.  As @DirtyHairy says, this is an idiosyncrasy of the hardware, and something that the debugger should know about.

What a wonderful beast the 2600 is, still causing problems after all these years :smiling_imp:

**Comment by thrust26** [comment]
@SvOlli Is "ghost read" the correct term here?

**Comment by thrust26** [comment]
Implemented option with 481c85c

**Comment by ale-79** [comment]
Related to this, I think that in the (rare) event of multiple traps triggered by the same instruction, the status line should indicate all of them. Currently if an instruction triggers two traps (with a "dummy" memory access and a "normal" one), only the latter is indicated (a breakpoint on the instruction following the one that triggered a trap isn't displayed either, but that's not a big deal maybe)

e.g. in this rom (that doesn't do anything useful other than displaying a yellow background...), try using trapreads on f135 and f235 (and a breakpoint at f03b)

[display.zip](https://github.com/stella-emu/stella/files/1569508/display.zip)

**Comment by thrust26** [comment]
That's a bit more complicated to achieve and I am not so convinced that those rare cases are worth the effort. 

Theoretically one instruction hit an unlimited account of traps and breakifs. And each peek and poke of the instruction can trigger those. I think currently the debugger only remembers the very last trigger (the previous ones get overwritten). 

To display them all, we would have to change the UI and display a list of triggers. And I am not sure that this display won't irritate more than it helps.

**Comment by ale-79** [comment]
You're right, I was thinking just about the extra dummy read and eventually a following break, but I forgot to consider the conditional traps and breaks...

**Comment by thrust26** [comment]
> I don't think this can be fixed before 5.1, though.

@sa666666 Wrong! 😈
